File: 2023/fsharp/Day07.fs
module Aoc.Solutions.Y2023.Day07

open Aoc.Runner
open System
open System.Text.RegularExpressions
open System.Collections.Generic

type Hand = list<int>
type Bid = (Hand * int)

type HandType =
| Five = 5
| Four = 4
| Three = 3
| TwoPair = 2
| OnePair = 1
| HighCard = 0

let handeType (hand: Hand) : HandType =
    let byNumber =
        hand
        |> Seq.groupBy (fun x -> x)
        |> Seq.map (fun x -> (fst x), (snd x ) |> Seq.length)
        |> Seq.sortByDescending snd
        |> List.ofSeq
    match (Seq.head byNumber |> snd) with
    | 5 -> HandType.Five
    | 4 -> HandType.Five
    | 3 -> HandType.Three
    | 2 ->
        match byNumber |> Seq.filter (fun z -> (fst z) = 2) |> Seq.length with
        | 2 -> HandType.TwoPair
        | _ -> HandType.OnePair
    | _ -> HandType.HighCard


let compareHand left right =
    if left = right then
        printfn $"{left} = {right}"
        -1 
    else
        let leftType = handeType left
        let rightType = handeType right
        if leftType <> rightType then
            leftType.CompareTo(rightType)
        else
            let (l, r) =
                seq {
                    for i in 0..Seq.length left do
                        let lv, rv  = left[i], right[i]
                        if lv <> rv then
                            yield lv, rv
                }
                |> Seq.head
                
            l.CompareTo(r)

let compareBid (left: Bid) (right: Bid) =
    compareHand (fst left) (fst right)

let parseLine (line: string) =
    let (left, right) = ((line.Split ' ')[0], (line.Split ' ')[1])
    let hand =
        left
        |> Seq.map (fun x ->
            match x with
            | 'A' -> 14
            | 'K' -> 13
            | 'Q' -> 12
            | 'J' -> 11
            | 'T' -> 10
            | _ -> int $"{x}")
            |> List.ofSeq
    (hand, int right): Bid

let parse (input: string) =
    lines input
    |> Seq.map parseLine
    |> List.ofSeq

type Day07() =
    inherit Day()

    override _.SolveA input =
        let bids = parse input |> ResizeArray
        
        let sorted =
            bids
            |> Seq.sortWith compareBid
            |> Seq.toList

        let ost =
            sorted
            |> Seq.fold (fun s x ->
                    let hand, bid = x
                    let line = $"{json hand} {bid} {handeType hand}"
                    $"{line}\n{s}"
                ) ""
        printfn $"{ost}"
        printfn $"{sorted |> Seq.map fst |> Seq.distinct |> Seq.length}"
        sorted
        |> Seq.indexed
        |> Seq.map (fun (i, bid) -> (int64 (snd bid)) * int64(i + 1))
        |> Seq.sum
        |> string


    override _.SolveB input =
        ""


    member _.samp = raw"""
    32T3K 765
    T55J5 684
    KK677 28
    KTJJT 220
    QQQJA 483
    """

    override this.Tests =
        [
            Test("a", this.samp, "6440", fun x -> this.SolveA x);
        ]
        |> List<Test>

File: 2023/csharp/Day07.cs
using System.Numerics;

namespace Aoc.Solutions.Y2023;

public class Day07 : Day
{
    enum HandType
    {
        Five = 5,
        Four = 4,
        Full = 3,
        Three = 2,
        TwoPair = 1,
        OnePair = 0,
        HighCard = -1,
    }

    record Bid(List<int> Hand, int Val) : IComparable<Bid>, IComparable
    {
        public override string ToString()
        {
            var toLetter = (int x) => x switch
            {
                14 => 'A',
                13 => 'K',
                12 => 'Q',
                11 => 'J',
                10 => 'T',
                _ => $"{x}"[0]
            };
            return $"{string.Join("", Hand.Select(toLetter))} {Val} ({Type()})";
        }
        public HandType Type()
        {
            var highest = Hand.GroupBy(x => x)
                .OrderByDescending(x => x.Count())
                .Select(x => x.Count())
                .First();
            var secondHighest = Hand.GroupBy(x => x)
                .OrderByDescending(x => x.Count())
                .Select(x => x.Count())
                .Skip(1)
                .FirstOrDefault();
            return highest switch
            {
                5 => HandType.Five,
                4 => HandType.Four,
                3 when secondHighest == 2 => HandType.Full,
                3 when secondHighest == 1 => HandType.Three,
                2 => secondHighest == highest ? HandType.TwoPair : HandType.OnePair,
                _ => HandType.HighCard
            };
        }

        public int CompareTo(Bid? other)
        {
            return CompareBid(this, other!);
        }

        public int CompareTo(object? obj)
        {
            if (obj is Bid b) CompareTo(b);
            return 0;
        }
    }
    
    Bid ParseLine(string line) =>
        new(
            line.Split(" ")[0].Select(c => "0123456789TJQKA".IndexOf(c)).ToList(),
            int.Parse(line.Split(" ")[1]));

    List<Bid> ParseInput(string input) =>
        input.Lines().Select(ParseLine).ToList();

    static int CompareBid(Bid left, Bid right)
    {
        if (left.Type() != right.Type())
            return left.Type().CompareTo(right.Type());
        return left.Hand.Zip(right.Hand).FirstOrDefault(x => x.First != x.Second)
            .Pipe(x => x.First.CompareTo(x.Second));
    }

    public override string SolveA(string input)
    {
        var parsed = input.Pipe(ParseInput).ToList();
        parsed.Sort();
        var d = parsed.ToDictionary(x => x.Val);
        return parsed.Select((x, i) => x.Val * (i + 1)).Sum().ToString();
    }

    public override string SolveB(string input)
    {
        throw new NotImplementedException();
    }

    public Day07()
    {
        Tests = new()
        {
            new("a", Sample, "6440", SolveA)
        };
    }

    string Sample = """
    32T3K 765
    T55J5 684
    KK677 28
    KTJJT 220
    QQQJA 483
    """;
}